cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -1000,26 +1001,42 @@ std::shared_ptr<std::unordered_map<std::pair<StreamId, VersionId>, AtomKey>> Loc
             return sym_versions.find(key) == std::end(sym_versions);
         });
         latest_versions = batch_get_latest_version(store(), version_map(), latest_ids, false);
-
     } else {
         specific_versions = std::make_shared<std::unordered_map<std::pair<StreamId, VersionId>, AtomKey>>();
         latest_versions = batch_get_latest_version(store(), version_map(), stream_ids, false);
     }
-    for(const auto& latest_version : *latest_versions)
-        specific_versions->try_emplace(std::make_pair(latest_version.first, latest_version.second.version_id()), latest_version.second);
 
-    return specific_versions;
+    return std::make_pair(specific_versions, latest_versions);
 }
 
 std::vector<std::pair<VariantKey, std::optional<google::protobuf::Any>>> LocalVersionedEngine::batch_read_metadata_internal(
     const std::vector<StreamId>& stream_ids,
     const std::vector<VersionQuery>& version_queries
     ) {
-    auto stream_index_map = get_stream_index_map(stream_ids, version_queries);
+    auto [specific_versions_index_map, latest_versions_index_map] = get_stream_index_map(stream_ids, version_queries);
+    auto version_queries_is_empty = version_queries.empty();
     std::vector<folly::Future<std::pair<VariantKey, std::optional<google::protobuf::Any>>>> fut_vec;
-    for (const auto &[_, key] : *stream_index_map)
-        fut_vec.push_back(store()->read_metadata(key));
-
+    for(const auto& stream_id : folly::enumerate(stream_ids)) {
+        auto has_specific_version_query = version_queries_is_empty ? 
+                                          false : std::holds_alternative<SpecificVersionQuery>(version_queries[stream_id.index].content_);
+        if(has_specific_version_query){
+            const auto& query = version_queries[stream_id.index].content_;
+            auto specific_version_map_key = std::make_pair(*stream_id, std::get<SpecificVersionQuery>(query).version_id_);
+            auto specific_version_it = specific_versions_index_map->find(specific_version_map_key);
+            if (specific_version_it != specific_versions_index_map->end()){
+                fut_vec.push_back(store()->read_metadata(specific_version_it->second));
+            }else{
+                fut_vec.push_back(std::make_pair(VariantKey(), std::nullopt));
```

Comment:
Don't create an empty `VariantKey`. It may cause issues in other C++ code calling this function.

Wrap the entire `pair` in an `optional` or not return the `VariantKey` at all.